presto/dataset.py

```diff
@@ -65,23 +65,73 @@ def target_crop(row_d: Dict) -> int:
         # by default, we predict crop vs non crop
         return int(row_d["LANDCOVER_LABEL"] == 11)
 
+    @classmethod
+    def get_timestep_positions(cls, row_d: Dict, augment: bool = False) -> List[int]:
+        available_timesteps = int(row_d["available_timesteps"])
+        valid_position = int(row_d["valid_position"])
+
+        # force moving the center point if it is too close to the edges
+        if (valid_position < cls.NUM_TIMESTEPS // 2) or (
+            valid_position > (available_timesteps - cls.NUM_TIMESTEPS // 2)
+        ):
+            augment = True
+
+        if not augment:
+            # Center the timesteps around the valid position
+            center_point = valid_position
+        else:
+            # Shift the center point but make sure the resulting range
+            # well includes the valid position
+
+            min_center_point = max(
+                cls.NUM_TIMESTEPS // 2, valid_position + MIN_EDGE_BUFFER - cls.NUM_TIMESTEPS // 2
+            )
+            max_center_point = min(
+                available_timesteps - cls.NUM_TIMESTEPS // 2,
+                valid_position - MIN_EDGE_BUFFER + cls.NUM_TIMESTEPS // 2,
+            )
+
+            center_point = np.random.randint(
+                min_center_point, max_center_point + 1
+            )  # max_center_point included
+
+        last_timestep = min(available_timesteps, center_point + cls.NUM_TIMESTEPS // 2)
+        first_timestep = max(0, last_timestep - cls.NUM_TIMESTEPS)
+        timestep_positions = list(range(first_timestep, last_timestep))
+
+        if len(timestep_positions) != cls.NUM_TIMESTEPS:
+            raise ValueError(
+                f"Acquired timestep positions do not have correct length: \
+required {cls.NUM_TIMESTEPS}, got {len(timestep_positions)}"
+            )
+        assert (
+            valid_position in timestep_positions
+        ), f"Valid position {valid_position} not in timestep positions {timestep_positions}"
+        return timestep_positions
+
     @classmethod
     def row_to_arrays(
-        cls, row: pd.Series, target_function: Callable[[Dict], int]
+        cls, row: pd.Series, target_function: Callable[[Dict], int], augment: bool = False
     ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, float, int]:
         # https://stackoverflow.com/questions/45783891/is-there-a-way-to-speed-up-the-pandas-getitem-getitem-axis-and-get-label
         # This is faster than indexing the series every time!
         row_d = pd.Series.to_dict(row)
 
         latlon = np.array([row_d["lat"], row_d["lon"]], dtype=np.float32)
-        month = datetime.strptime(row_d["start_date"], "%Y-%m-%d").month - 1
+
+        timestep_positions = cls.get_timestep_positions(row_d, augment=augment)
+        # make sure that month for encoding gets shifted according to
+        # the selected timestep positions
+        month = (
+            pd.to_datetime(row_d["start_date"]) + pd.DateOffset(months=timestep_positions[0])
```

Comment:
You could also just add `timestep_positions[0]` and not use the overhead of the `pd.DateOffset` method. I don't think we're resilient in any case to situations where we're working with dekadal data. This is a potential risk if someone (including ourselves) is going over the dekadal track. Can we make it more universal here? Or if needed for now do a check and raise something, so nobody blindly takes this while it won't work as expected.

Reply:
good point. tried to account for that here ad05b3de8acbd5379e776fdb92969e484bd1da3c
we will also need to think about renaming this `month` thing (since it can be something else too), and also making sure that other relative timestep positions (`valid_position` and `timestep_ind`) are computed not as month, but in a more generic fashion

Reply:
You mean this has to be tackled in `process_parquet`?

Reply:
I am not sure I understand what `initial_start_date_position` actually means and why it cannot just be the month inferred from `start_date`.

Reply:
well, maybe I'm overthinking it...

here's an example:
1. we are working on a monthly basis, with start_date in October (hence, `month = 10`). 
and we want to shift it 4 timesteps forward. we can just make `(10 + 4) - 1` and add the modulo part not to get bad month value `% cls.NUM_TIMESTEPS`

2. we are working on a dekadal basis. start date is the same. I assume that our timestep indices are not in month chunks, but in 10-day intervals. so, we have observations for ts0, ts1, ..., ts45 (for example). our `NUM_TIMESTEPS` variable should be set to 36 instead of 12 (like Giorgia did). the `valid_date` should also translate into 10-day chunk instead of month. so that we can select 36 timesteps around the `valid_position`. 
now, we have our `timestep_positions` returned in dekadal steps. so when we add 4 dekadal steps to a month value, it doesn't make sense. we need to add apples to apples to get the date that accounts for the shift.
and now we can take it's month and pass it to presto. I realize now that this particular step is missing in my implementation.

am I missing something?

Reply:
Following most of it. But still, `start_date` is a real date, from which we can infer what would normally be the start month being fed to Presto. Why do we need to translate it to a position? I agree that what we have to add to it due to the shift should account for the time "resolution". `valid_position` should be the translation of `valid_date` (which is irrespective of time resolution) to the position in the timesteps (which is depending on the time resolution). I might be overthinking it just as much.

Reply:
okay, maybe something like this can work:
```
step_converted_to_month = np.ceil(timestep_positions[0] * (365 // NUM_TIMESTEPS) / 30)
month = (pd.to_datetime(start_date).month + step_converted_to_month) % 12 - 1
```
it can be a little imprecise in some cases, not more than 1 month error.
but it's just a one-liner, so probably we can sacrifice this little bit of precision

Reply:
yeah let's make a note that Giorgia can test this out properly when doing dekadal runs.